### PR TITLE
config: Interpolate fields tagged with interpolate

### PR DESCRIPTION
**Pull request by abhinav**
This adds supports to x/config for including values from environment
variables while decoding if the target struct field is tagged with the
`interpolate` option.

In,

    type config struct {
        Address string `config:",interpolate"`
    }

Strings in the form `${FOO}` or `${FOO:bar}` will be interpolated using
internal/interpolate with environment variables being used to resolve
variables.

Depends on #942

### CODE REVIEW

**Comment by alsamylkin** [comment]
Could you shed more light, why a compile time interpolation flag is better than a runtime interpolation in decoder? It looks a bit restrictive, because e.g. map values can't be interpolated in the same way.

**Comment by abhinav** [comment]
@alsamylkin My bad. @kriskowal and I discussed this in-person but we failed to
document it anywhere. We don't want to apply the interpolation to all strings
because,

-   Certain strings (randomly generated keys, for example) can have the string
    `${` in them.
-   We intend to support template-like strings for routing in the YAML.
    Imagine, `/foo/${dc}` where `${dc}` is not filled by an environment
    variable but some other source controlled by the transport author.
    Replacing all variables will break that in a bad way.
-   We foresee no compelling use case for interpolation in map keys.

Based on that, we decided that it was best for interpolation to be opt-in on
specific fields. The transport author will decide which fields support
interpolation from environment variables.

Re: Map values:

Are you talking about,

    # FOO="{port: 80}"
    someMap: ${FOO}

Or,

    # FOO=80
    someMap:
      foo: ${FOO}

If it's the former, we have no plans of supporting that at
this time. We don't think there are compelling use cases for holding complex
values in single environment variables and would like to discourage that
pattern.

If it's the latter, yeah you're right. The initial implementation (in this PR)
only fills fields tagged with `interpolate` that are parseable from a string
(primitives and `time.Duration`).  A follow up PR can add support for
intepolating values inside maps and lists so you should be able to do,

    type myConfig struct {
      SomeList []string `config:",interpolate"`
      SomeMap map[string]string `config:",interpolate"`
    }

    /* And then,
    
    someList:
      - foo
      - ${BAR}
      - baz
    someMap:
      foo: ${BAR}
    */

**Comment by alsamylkin** [comment]
Yeah, I meant the latter, there are also possibilities that keys can contain interpolation parameters:
```
routingMap:
  some${dc}:${owner}
```
where you may want to have different interpolation functions, etc.

How do you see a configuration with two lookup functions working together(one `os.LookupEnv` and one is something else)?

File: x/config/mapdecode.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+		// have received an integer as expected.
+		return data, nil
+	}
+
+	s, err := interpolate.Parse(v)
+	if err != nil {
+		return data, fmt.Errorf("failed to parse %q for interpolation: %v", v, err)
+	}
+
+	newV, err := s.Render(os.LookupEnv)
```
Comment (alsamylkin) [inline x/config/mapdecode.go]: Nit: could you pull up the lookUp function to be a public parameter? Testing will be much easier for users, I mean writing parallel tests mostly.

**Comment by abhinav** [comment]
For the keys, the consensus was that we don't need it that strongly. If we
find that we need it in the future, it can be added but it's not worth the
effort right now IMO.

Instead of two lookup functions, the idea is that the transport author will
just state that the attribute should not be interpolated and call the
interpolation library manually after decoding. So this depends on us making
the interpolate package available separately.

File: x/config/mapdecode.go (hunk 1 above)
Comment (abhinav) [inline x/config/mapdecode.go]: I think I'll do this in a separate PR because this requires adding options to `configurator.New`.

File: x/config/configurator_test.go [hunk 2]
```diff
@@ -854,6 +1101,11 @@ func TestConfigurator(t *testing.T) {
 			tt := tc.test(t, mockCtrl)
 			cfg := New()
 
+			for key, value := range tt.env {
+				restore := setenv(t, key, value)
```
Comment (alsamylkin) [inline x/config/configurator_test.go]: why not `defer setenv(t, key, value)()`?

Comment (abhinav) [inline x/config/configurator_test.go]: It's the same but I like the explicit variable in this case. Evaluation order (the fact that `setenv(..)` is called right away) can be confusing for newcomers.

File: x/config/configurator_test.go [hunk 3]
```diff
@@ -834,6 +850,237 @@ func TestConfigurator(t *testing.T) {
 				return
 			},
 		},
+		{
```
Comment (prashantv) [inline x/config/configurator_test.go]: suggest a couple of negative tests:
 - fields without `interpolate` that contain interpolation strings like `${REDIS_PORT}` are not interpolated
 - map keys/values and slice values are not interpolated

Comment (prashantv) [inline x/config/configurator_test.go]: Also having a couple of these as integration tests that check the final transport will be useful, but all the interpolation test cases as integration tests might be a little heavy. consider just having unit tests in a `mapdecode_test.go` that tests all the corner cases of interpolation (like what is interpolated, what isn't). they can just check the final struct rather than building a config out of the structs.

File: x/config/mapdecode.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+func decodeInto(dst, src interface{}) error {
+	return mapdecode.Decode(dst, src,
+		mapdecode.TagName(_tagName),
+		mapdecode.FieldHook(interpolateHook))
+}
+
+func interpolateHook(from reflect.Type, to reflect.StructField, data reflect.Value) (reflect.Value, error) {
+	shouldInterpolate := false
+
+	options := strings.Split(to.Tag.Get(_tagName), ",")[1:]
```
Comment (willhug) [inline x/config/mapdecode.go]: assuming we can grab any tag here, why are we adding interpolation to the `config` tag which is already used for specifying the name of the field, can we just add another tag?

Comment (abhinav) [inline x/config/mapdecode.go]: Instead of reserving another tag name, this allows us to just use the existing tag. It's similar to the json/yaml pattern where they have the name and then a comma-separated list of options.

    json:"foo,omitempy"

Tags usually have to have a value. I'm not sure what the value for `interpolate` would be.

File: x/config/mapdecode.go (hunk 1 above)
Comment (abhinav) [inline x/config/mapdecode.go]: @alsamylkin Follow-up change here: #953
